spectec/spec/6-reduction.watsup

```diff
@@ -172,44 +173,46 @@ rule Step_pure/local.tee:
 
 
 rule Step_read/global.get:
-  z; (GLOBAL.GET x)  ~>  $global(z, x)
+  z; (GLOBAL.GET x)  ~>  $global(z, x).VALUE
 
 rule Step/global.set:
   z; val (GLOBAL.SET x)  ~>  $with_global(z, x, val); epsilon
 
 
 rule Step_read/table.get-trap:
   z; (CONST I32 i) (TABLE.GET x)  ~>  TRAP
-  -- if i >= |$table(z, x)|
+  -- if i >= |$table(z, x).ELEM|
 
 rule Step_read/table.get-val:
-  z; (CONST I32 i) (TABLE.GET x)  ~>  $table(z,x)[i]
-  -- if i < |$table(z, x)|
+  z; (CONST I32 i) (TABLE.GET x)  ~>  $table(z,x).ELEM[i]
+  -- if i < |$table(z, x).ELEM|
 
 rule Step/table.set-trap:
   z; (CONST I32 i) ref (TABLE.GET x)  ~>  z; TRAP
-  -- if i >= |$table(z, x)|
+  -- if i >= |$table(z, x).ELEM|
 
 rule Step/table.set-val:
   z; (CONST I32 i) ref (TABLE.GET x)  ~>  $with_table(z, x, i, ref); epsilon
-  -- if i < |$table(z, x)|
+  -- if i < |$table(z, x).ELEM|
 
 
 rule Step_read/table.size:
   z; (TABLE.SIZE x)  ~>  (CONST I32 n)
-  -- if |$table(z, x)| = n  ;; TODO: inline this
+  -- if |$table(z, x).ELEM| = n  ;; TODO: inline this
 
 
 rule Step/table.grow-succeed:
-  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableext(z, x, ref^n); (CONST I32 $(|$table(z, x)|))
+  z; ref (CONST I32 n) (TABLE.GROW x)  ~>  $with_tableext(z, x, ref^n); (CONST I32 $(|$table(z, x).ELEM|))
+  -- if $table(z, 0).TYPE = tabletype
+  -- Tabletype_ok: |- tabletype : OK
```

Comment:
Hm, AFAICS, that would just revalidate the existing table type, which should already be valid.

What is needed here are three steps: (1) modifying the table type to the new lower limit, (2) checking that these new limits are still valid, if okay, update the table's actual type to the new type.

Reply:
Oh, thank you for the comment. I fixed it.

Reply:
I'm afraid this hasn't fixed it, it's still just revalidating the old type unchanged.

Growing a table/memory updates the lower limit of its instance's dynamic type to the new size. And the validity check on the type is one way to ensure that this doesn't violate the upper limit.